src/axom/core/Array.hpp

```diff
@@ -550,121 +827,291 @@ Array< T >::~Array()
+  return first - shifted;
 }
 
 //------------------------------------------------------------------------------
 template< typename T >
-inline void Array< T >::insert( const T* tuples, IndexType n, IndexType pos )
+template< typename ... Args >
+inline void Array< T >::emplace( IndexType pos, Args&&... args )
 {
-  assert( tuples != nullptr );
-  T* insert_pos = reserveForInsert( n, pos );
-  std::memcpy( insert_pos, tuples, n * m_num_components * sizeof(T) );
+  reserveForInsert( 1, pos );
+  m_data[ pos ] = std::move( T(std::forward<Args>(args)...));
 }
 
 //------------------------------------------------------------------------------
 template< typename T >
-inline void Array< T >::emplace( IndexType n, IndexType pos, const T& value )
+template< typename ... Args >
+inline typename Array< T >::ArrayIterator
+Array< T >::emplace( Array< T >::ArrayIterator pos, Args&&... args )
 {
-  T* insert_pos = reserveForInsert( n, pos );
-  std::fill_n( insert_pos, n * numComponents(), value );
+  assert( pos >= begin() && pos <= end() );
+  emplace( pos - begin(), args... );
+  return pos;
 }
 
+//------------------------------------------------------------------------------
+template< typename T >
+template< typename ... Args >
+inline void Array< T >::emplace_back( Args&&... args )
+{
+  emplace( m_num_elements, args... );
+}
 
 //------------------------------------------------------------------------------
 template< typename T >
-inline void Array< T >::resize( IndexType new_num_tuples )
+inline void Array< T >::resize( IndexType new_num_elements )
 {
-  assert( new_num_tuples >= 0 );
+  assert( new_num_elements >= 0 );
 
-  if ( new_num_tuples > m_capacity )
+  if ( new_num_elements > m_capacity )
   {
-    dynamicRealloc( new_num_tuples );
+    dynamicRealloc( new_num_elements );
   }
 
-  updateNumTuples( new_num_tuples );
+  updateNumElements( new_num_elements );
+}
+
+//------------------------------------------------------------------------------
+template< typename T >
+inline void Array< T >::swap ( Array< T >& other )
+{
+  T* temp_data = m_data;
+  IndexType temp_num_elements = m_num_elements;
+  IndexType temp_capacity = m_capacity;
+  double temp_resize_ratio = m_resize_ratio;
+  bool temp_is_external = m_is_external;
+
+  m_data = other.m_data;
+  m_num_elements = other.m_num_elements;
+  m_capacity = other.m_capacity;
+  m_resize_ratio = other.m_resize_ratio;
+  m_is_external = other.m_is_external;
+
+  other.m_data = temp_data;
+  other.m_num_elements = temp_num_elements;
+  other.m_capacity = temp_capacity;
+  other.m_resize_ratio = temp_resize_ratio;
+  other.m_is_external = temp_is_external;
+
 }
 
 //------------------------------------------------------------------------------
 template< typename T >
-inline void Array< T >::initialize( IndexType num_tuples,
-                                    IndexType num_components,
+inline std::ostream& Array< T >::print(std::ostream& os) const
+{
```

Comment:
Make sure we copy the array on the host? 

If the Array is on the GPU and this gets called, perhaps we could throw a warning:
```
"You are printing an Array whose contents are on the GPU. I hope you know what you're doing!"
```